Reliabledatachannelfixes

Terricide opened the pull request.

sipsorcery commented on src/net/SCTP/Association.cs:
```diff
@@ -37,6 +37,7 @@ namespace SIPSorcery.Net.Sctp
 {
     public abstract class Association
     {
+        protected object myLock = new object();
```
I don't see this used anywhere? Is it needed?

Terricide replied:
No it isn't. I was planning on another pull request to change instances of lock(this) since it isnt recommend.

https://stackoverflow.com/questions/11788582/best-practices-in-using-a-lock

sipsorcery commented:
@Terricide do you get similar output to below when you run the [Data Channel Stress Test example app](https://github.com/sipsorcery/sipsorcery/tree/master/examples/WebRTCScenarios/DataChannelStressTest)? 

````
examples\WebRTCScenarios\DataChannelStressTest> dotnet run
....
[11:38:58 DBG] Data channel for label dcx now open.
Data channel send 0 on dcx.
Data channel send 1 on dcx.
Data channel send 2 on dcx.
Data channel send 3 on dcx.
Data channel send 4 on dcx.
Data channel send 5 on dcx.
Data channel send 6 on dcx.
Data channel send 7 on dcx.
Data channel send 8 on dcx.
Data channel send 9 on dcx.
Data channel send 10 on dcx.
Data channel send 11 on dcx.
Data channel send 12 on dcx.
[11:38:59 DBG] PeerB: Data channel receive: 2, length 4.
Data channel send 13 on dcx.
[Data channel send 14 on dcx.
11:38:59Data channel send 15 on dcx.
 Data channel send 16 on dcx.
Data channel send 17 on dcx.
DBGData channel send 18 on dcx.
] Data channel send 19 on dcx.
PeerB: Data channel receive: 6, length 4.Data channel send 20 on dcx.

Data channel send 21 on dcx.
Data channel send 22 on dcx.
[Data channel send 23 on dcx.
11:38:59Data channel send 24 on dcx.
 Data channel send 25 on dcx.
DBGData channel send 26 on dcx.
] Data channel send 27 on dcx.
PeerB: Data channel receive: 5, length 4.Data channel send 28 on dcx.

Data channel send 29 on dcx.
[11:38:59 DBG] PeerB: Data channel receive: 3, length 4.
[11:38:59 DBG] PeerB: Data channel receive: 11, length 4.
[11:38:59 DBG] PeerB: Data channel receive: 12, length 4.
[11:38:59 DBG] PeerB: Data channel receive: 4, length 4.
[11:38:59 DBG] PeerB: Data channel receive: 7, length 4.
[11:38:59 DBG] PeerB: Data channel receive: 16, length 4.
[11:38:59 DBG] PeerB: Data channel receive: 17, length 4.
[11:38:59 DBG] PeerB: Data channel receive: 10, length 4.
[11:38:59 DBG] PeerB: Data channel receive: 19, length 4.
[11:38:59 DBG] PeerB: Data channel receive: 20, length 4.
[11:38:59 DBG] PeerB: Data channel receive: 13, length 4.
[11:38:59 DBG] PeerB: Data channel receive: 14, length 4.
[11:38:59 DBG] PeerB: Data channel receive: 15, length 4.
[11:38:59 DBG] PeerB: Data channel receive: 24, length 4.
[11:38:59 DBG] PeerB: Data channel receive: 9, length 4.
[11:38:59 DBG] PeerB: Data channel receive: 18, length 4.
[11:38:59 DBG] PeerB: Data channel receive: 1, length 4.
[11:38:59 DBG] PeerB: Data channel receive: 28, length 4.
[11:38:59 DBG] PeerB: Data channel receive: 29, length 4.
[11:38:59 DBG] PeerB: Data channel receive: 22, length 4.
[11:38:59 DBG] PeerB: Data channel receive: 23, length 4.
[11:38:59 DBG] PeerB: Data channel receive: 8, length 4.
[11:38:59 DBG] PeerB: Data channel receive: 25, length 4.
[11:38:59 DBG] PeerB: Data channel receive: 26, length 4.
[11:38:59 DBG] PeerB: Data channel receive: 27, length 4.
[11:38:59 DBG] PeerB: Data channel receive: 0, length 4.
[11:38:59 DBG] PeerB: Data channel receive: 21, length 4.
````

Terricide commented:
I took off ThreadPool.QueueUserWorkItem in the BlockingSCTPStream class. This fixes the messages arriving in the right order for me. 
I think the user can spawn threads as messages arrive if they need to do some heavy processing. I also added back the message ordering checks I removed since it seems to be working fine now with the BlockingCollection changes I made.

sipsorcery commented:
Thanks, looks good now.

````
[14:01:08 DBG] Data channel for label dcx now open.
Data channel send 0 on dcx.
Data channel send 1 on dcx.
Data channel send 2 on dcx.
Data channel send 3 on dcx.
Data channel send 4 on dcx.
Data channel send 5 on dcx.
Data channel send 6 on dcx.
Data channel send 7 on dcx.
Data channel send 8 on dcx.
Data channel send 9 on dcx.
Data channel send 10 on dcx.
Data channel send 11 on dcx.
[Data channel send 12 on dcx.
Data channel send 13 on dcx.
14:01:09Data channel send 14 on dcx.
 Data channel send 15 on dcx.
DBGData channel send 16 on dcx.
] Data channel send 17 on dcx.
PeerB: Data channel receive: 0, length 4.Data channel send 18 on dcx.

Data channel send 19 on dcx.
Data channel send 20 on dcx.
Data channel send 21 on dcx.
Data channel send 22 on dcx.
[14:01:09 DBG] PeerB: Data channel receive: 1, length 4.
[14:01:09 DBGData channel send 23 on dcx.
] Data channel send 24 on dcx.
PeerB: Data channel receive: 2, length 4.Data channel send 25 on dcx.

Data channel send 26 on dcx.
Data channel send 27 on dcx.
[14:01:09 DBG] PeerB: Data channel receive: 3, length 4.
[14:01:09 DBG] PeerB: Data channel receive: 4, length 4.Data channel send 28 on dcx.

Data channel send 29 on dcx.
[14:01:09 DBG] PeerB: Data channel receive: 5, length 4.
[14:01:09 DBG] PeerB: Data channel receive: 6, length 4.
[14:01:09 DBG] PeerB: Data channel receive: 7, length 4.
[14:01:09 DBG] PeerB: Data channel receive: 8, length 4.
[14:01:09 DBG] PeerB: Data channel receive: 9, length 4.
[14:01:09 DBG] PeerB: Data channel receive: 10, length 4.
[14:01:09 DBG] PeerB: Data channel receive: 11, length 4.
[14:01:09 DBG] PeerB: Data channel receive: 12, length 4.
[14:01:09 DBG] PeerB: Data channel receive: 13, length 4.
[14:01:09 DBG] PeerB: Data channel receive: 14, length 4.
[14:01:09 DBG] PeerB: Data channel receive: 15, length 4.
[14:01:09 DBG] PeerB: Data channel receive: 16, length 4.
[14:01:09 DBG] PeerB: Data channel receive: 17, length 4.
[14:01:09 DBG] PeerB: Data channel receive: 18, length 4.
[14:01:09 DBG] PeerB: Data channel receive: 19, length 4.
[14:01:09 DBG] PeerB: Data channel receive: 20, length 4.
[14:01:09 DBG] PeerB: Data channel receive: 21, length 4.
[14:01:09 DBG] PeerB: Data channel receive: 22, length 4.
[14:01:09 DBG] PeerB: Data channel receive: 23, length 4.
[14:01:09 DBG] PeerB: Data channel receive: 24, length 4.
[14:01:09 DBG] PeerB: Data channel receive: 25, length 4.
[14:01:09 DBG] PeerB: Data channel receive: 26, length 4.
[14:01:09 DBG] PeerB: Data channel receive: 27, length 4.
[14:01:09 DBG] PeerB: Data channel receive: 28, length 4.
[14:01:09 DBG] PeerB: Data channel receive: 29, length 4.
````
